File: backend/internal/demo/fleet_test.go
package demo_test

import (
	"encoding/json"
	"testing"
	"time"

	"github.com/Alisher24/CarSharing/backend/internal/demo"
	"github.com/Alisher24/CarSharing/backend/internal/fleet"
	"github.com/Alisher24/CarSharing/backend/internal/rentals/stage"
)

// declaredFleetSize is the size the demonstration promises: five vehicles of each of the five
// powertrains.
const declaredFleetSize = 25

// observedAt stands for the moment the catalog is read. Every declared vehicle has just confirmed
// its position when the scenario is installed, so freshness is not what these cases are about.
var observedAt = time.Date(2026, time.September, 13, 7, 15, 30, 0, time.UTC)

// published resolves a declared vehicle to the state the catalog would show for it.
func published(vehicle demo.Vehicle) fleet.State {
	return fleet.Vehicle{
		PowertrainType: vehicle.PowertrainType,
		Connected:      vehicle.Connected,
		Telemetry:      fleet.Telemetry{Position: vehicle.Position, ConfirmedAt: observedAt},
		ServiceZoneID:  "01994342-6ba7-7000-8000-000200000001",
		Sources:        vehicle.Sources,
		HeldBy:         vehicle.HeldBy,
	}.StateAt(observedAt)
}

func TestEveryPowertrainShowsEveryPublicState(t *testing.T) {
	declared := demo.Fleet()
	if len(declared) != declaredFleetSize {
		t.Fatalf("fleet has %d vehicles, want %d", len(declared), declaredFleetSize)
	}

	wanted := map[fleet.Status]int{
		fleet.Available:   2,
		fleet.Reserved:    1,
		fleet.InTrip:      1,
		fleet.Unavailable: 1,
	}
	for _, powertrain := range fleet.PowertrainTypes() {
		t.Run(string(powertrain), func(t *testing.T) {
			counted := map[fleet.Status]int{}
			for _, vehicle := range declared {
				if vehicle.PowertrainType == powertrain {
					counted[published(vehicle).Status]++
				}
			}
			for status, count := range wanted {
				if counted[status] != count {
					t.Errorf("%q appears %d times, want %d", status, counted[status], count)
				}
			}
		})
	}
}

func TestPreparedTripsShowBothDrivingAndPausing(t *testing.T) {
	modes := map[fleet.RideMode]int{}
	for _, vehicle := range demo.Fleet() {
		if state := published(vehicle); state.Status == fleet.InTrip {
			modes[state.RideMode]++
		}
	}
	if modes[fleet.Driving] == 0 || modes[fleet.Paused] == 0 {
		t.Fatalf("prepared trips are %v, want both driving and paused", modes)
	}
}

func TestEveryPowertrainKeepsAnExhaustedExample(t *testing.T) {
	for _, powertrain := range fleet.PowertrainTypes() {
		t.Run(string(powertrain), func(t *testing.T) {
			for _, vehicle := range demo.Fleet() {
				if vehicle.PowertrainType != powertrain {
					continue
				}
				state := published(vehicle)
				if state.Status == fleet.Unavailable && hasReason(state, fleet.InsufficientEnergy) {
					return
				}
			}
			t.Fatal("no vehicle of this powertrain stands as an exhausted example")
		})
	}
}

func TestExhaustedExamplesAlsoShowStaleAndOfflineVehicles(t *testing.T) {
	var stale, offline int
	for _, vehicle := range demo.Fleet() {
		switch {
		case !vehicle.Connected:
			offline++
		case !vehicle.Reporting:
			stale++
		}
	}
	if stale == 0 || offline == 0 {
		t.Fatalf("%d vehicles go stale and %d are unlinked, want at least one of each", stale, offline)
	}
}

// A vehicle a person may book by hand must be left alone by the scenario command, and every
// vehicle the command does put back must be one the demonstration itself set up.
func TestRestorationCoversPreparedVehiclesOnly(t *testing.T) {
	for _, vehicle := range demo.Fleet() {
		free := published(vehicle).Status == fleet.Available
		if free == vehicle.Restored() {
			t.Errorf("%s: available %t, restored %t", vehicle.Model, free, vehicle.Restored())
		}
	}
}

// A prepared rental belongs to a service account of its own, because one account may hold only one
// live rental and none of them may be an account a person signs in as.
func TestEveryPreparedRentalHasItsOwnServiceAccount(t *testing.T) {
	accounts := map[string]string{}
	for _, vehicle := range demo.Fleet() {
		if vehicle.HeldBy == stage.NotHeld {
			if vehicle.ScenarioAccount != "" || vehicle.PreparedRentalID != "" {
				t.Errorf("%s has no prepared rental but names one", vehicle.Model)
			}
			continue
		}
		if previous, taken := accounts[vehicle.ScenarioAccount]; taken {
			t.Errorf("%s shares an account with %s", vehicle.Model, previous)
		}
		accounts[vehicle.ScenarioAccount] = vehicle.Model
	}
	for _, manual := range demo.ManualCheckAccounts {
		if held, taken := accounts[manual]; taken {
			t.Errorf("a manual check account holds the rental of %s", held)
		}
	}
}

// A vehicle parked outside the boundary would be published as outside the service zone and would
// stop being the example it was declared to be, so every declared position is checked against the
// area the command actually installs rather than against a copy of its corners. That area is a
// rectangle, so its bounding box is the area itself.
func TestEveryDeclaredVehicleStandsInsideTheServiceZone(t *testing.T) {
	west, south, east, north := installedAreaBounds(t)
	for _, vehicle := range demo.Fleet() {
		position := vehicle.Position
		inside := position.Longitude >= west && position.Longitude <= east &&
			position.Latitude >= south && position.Latitude <= north
		if !inside {
			t.Errorf("%s stands at %v, %v: outside the installed area",
				vehicle.Model, position.Longitude, position.Latitude)
		}
	}
}

func installedAreaBounds(t *testing.T) (west, south, east, north float64) {
	t.Helper()
	var area struct {
		Coordinates [][][2]float64 `json:"coordinates"`
	}
	if err := json.Unmarshal(demo.Zone().Area, &area); err != nil {
		t.Fatalf("the installed area is not readable as GeoJSON: %v", err)
	}
	if len(area.Coordinates) != 1 || len(area.Coordinates[0]) < 3 {
		t.Fatalf("the installed area is %d rings, want one ring of at least three points", len(area.Coordinates))
	}
	ring := area.Coordinates[0]
	west, south, east, north = ring[0][0], ring[0][1], ring[0][0], ring[0][1]
	for _, corner := range ring {
		west, east = min(west, corner[0]), max(east, corner[0])
		south, north = min(south, corner[1]), max(north, corner[1])
	}
	return west, south, east, north
}

func hasReason(state fleet.State, wanted fleet.UnavailableReason) bool {
	for _, reason := range state.UnavailableReasons {
		if reason == wanted {
			return true
		}
	}
	return false
}

File: scripts/acceptance/fleet.test.mjs
// The public catalog, the service zone and the tariff, observed on the real HTTP boundary and in
// the PostGIS the running service reads. Every claim here is made about the assembled stack: the
// fleet these tests describe is the one a visitor loads in a browser.
import assert from 'node:assert/strict';
import { before, describe, test } from 'node:test';
import { call, sql, waitForReady } from './client.mjs';
import {
  countStates,
  DRIVING_RATE_TYIYN,
  everySource,
  FLEET_SIZE,
  PAUSED_RATE_TYIYN,
  pointAt,
  POWERTRAIN_TYPES,
  restoreScenario,
  runSeed,
  scalar,
  START_THRESHOLD_BASIS_POINTS,
  STATES_PER_POWERTRAIN,
  TARIFFS_PATH,
  VEHICLES_PATH,
  vehiclesOfPowertrain,
  ZONES_PATH,
} from './fleet.mjs';

/** Words a public answer must never contain, because none of them is anybody's business. */
const PRIVATE_WORDS = ['user', 'email', 'rental', 'invoice', 'route', 'history'];

async function readCatalog() {
  const response = await call(VEHICLES_PATH);
  assert.equal(response.status, 200, response.text);
  return response.json;
}

before(async () => {
  await waitForReady();
  runSeed();
});

describe('the demonstration fleet a visitor loads', () => {
  test('holds five vehicles of each of the five powertrains', async () => {
    const catalog = await readCatalog();
    assert.equal(catalog.items.length, FLEET_SIZE);
    for (const powertrain of POWERTRAIN_TYPES) {
      assert.equal(vehiclesOfPowertrain(catalog.items, powertrain).length, 5, powertrain);
    }
  });

  test('shows every public state within every powertrain', async () => {
    const catalog = await readCatalog();
    for (const powertrain of POWERTRAIN_TYPES) {
      const counted = countStates(vehiclesOfPowertrain(catalog.items, powertrain));
      assert.deepEqual(counted, STATES_PER_POWERTRAIN, powertrain);
    }
  });

  test('shows both a moving and a paused ride among the prepared trips', async () => {
    const catalog = await readCatalog();
    const modes = catalog.items.filter((vehicle) => vehicle.status === 'in_trip').map((one) => one.ride_mode);
    assert.ok(modes.includes('driving'), `ride modes were ${modes.join(', ')}`);
    assert.ok(modes.includes('paused'), `ride modes were ${modes.join(', ')}`);
  });

  test('keeps a stable order between two readings', async () => {
    const first = await readCatalog();
    const second = await readCatalog();
    assert.deepEqual(
      second.items.map((vehicle) => vehicle.id),
      first.items.map((vehicle) => vehicle.id),
    );
  });

  test('lists every energy source separately, never a total', async () => {
    const catalog = await readCatalog();
    const hybrids = vehiclesOfPowertrain(catalog.items, 'hybrid');
    for (const hybrid of hybrids) {
      const kinds = hybrid.energy_sources.map((source) => source.kind).sort();
      assert.deepEqual(kinds, ['battery', 'gasoline'], hybrid.model);
    }
  });
});

describe('what makes a vehicle fit to start', () => {
  test('a source is startable exactly from the threshold upwards', async () => {
    const catalog = await readCatalog();
    for (const source of everySource(catalog.items)) {
      assert.equal(
        source.can_start,
        source.remaining_basis_points >= START_THRESHOLD_BASIS_POINTS,
        `${source.kind} at ${source.remaining_basis_points}`,
      );
    }
  });

  test('the fleet actually contains the two reserves the rule turns on', async () => {
    const catalog = await readCatalog();
    const shares = everySource(catalog.items).map((source) => source.remaining_basis_points);
    assert.ok(shares.includes(START_THRESHOLD_BASIS_POINTS), 'no source sits exactly at the threshold');
    assert.ok(
      shares.includes(START_THRESHOLD_BASIS_POINTS - 1),
      'no source sits one ten-thousandth below the threshold',
    );
  });

  test('reserves of two sources are not added together', async () => {
    const catalog = await readCatalog();
    const notSummed = catalog.items.filter(
      (vehicle) =>
        vehicle.energy_sources.length > 1 &&
        vehicle.energy_sources.every((source) => !source.can_start) &&
        vehicle.energy_sources.reduce((total, source) => total + source.remaining_basis_points, 0) >=
          START_THRESHOLD_BASIS_POINTS,
    );
    assert.ok(notSummed.length > 0, 'the fleet has no vehicle whose reserves would pass if added');
    for (const vehicle of notSummed) {
      assert.equal(vehicle.status, 'unavailable', vehicle.model);
      assert.ok(vehicle.unavailable_reasons.includes('insufficient_energy'), vehicle.model);
    }
  });

  test('an empty battery does not hide a sufficient tank', async () => {
    const catalog = await readCatalog();
    const emptyBattery = vehiclesOfPowertrain(catalog.items, 'hybrid').filter((vehicle) =>
      vehicle.energy_sources.some((source) => source.kind === 'battery' && source.remaining_basis_points === 0),
    );
    assert.ok(emptyBattery.length > 0, 'the fleet has no hybrid with an empty battery');
    for (const vehicle of emptyBattery) {
      assert.equal(vehicle.status, 'available', vehicle.model);
    }
  });

  test('every powertrain keeps an exhausted example', async () => {
    const catalog = await readCatalog();
    for (const powertrain of POWERTRAIN_TYPES) {
      const exhausted = vehiclesOfPowertrain(catalog.items, powertrain).filter(
        (vehicle) => vehicle.status === 'unavailable' && vehicle.unavailable_reasons.includes('insufficient_energy'),
      );
      assert.ok(exhausted.length > 0, powertrain);
    }
  });
});

describe('a rental decides occupancy', () => {
  test('a vehicle a rental holds is occupied whatever its own condition says', async () => {
    const catalog = await readCatalog();
    const held = sql(`SELECT vehicle_id FROM rentals WHERE ended_at IS NULL ORDER BY vehicle_id`).split('\n');
    for (const vehicle of catalog.items) {
      const occupied = vehicle.status === 'reserved' || vehicle.status === 'in_trip';
      assert.equal(occupied, held.includes(vehicle.id), `${vehicle.model} is ${vehicle.status}`);
    }
  });

  // Draining a rented vehicle and reading the catalog again is the only way to see that the rental
  // still decides: a low reserve on a free vehicle would make it unavailable, and on a rented one
  // it must change nothing. The scenario is put back afterwards.
  test('a rented vehicle that runs out of energy is still shown as rented', async () => {
    const catalog = await readCatalog();
    const reserved = catalog.items.find((vehicle) => vehicle.status === 'reserved');
    assert.ok(reserved, 'the scenario has no reserved vehicle');

    sql(`UPDATE vehicle_energy_sources SET remaining = 0 WHERE vehicle_id = '${reserved.id}'`);
    try {
      const drained = await call(`${VEHICLES_PATH}/${reserved.id}`);
      assert.equal(drained.status, 200, drained.text);
      assert.equal(drained.json.status, 'reserved');
      assert.ok(drained.json.energy_sources.every((source) => !source.can_start));
    } finally {
      restoreScenario();
    }
  });
});

describe('what a public answer may say', () => {
  test('the catalog, the zone and the tariff answer a reader with no account', async () => {
    for (const path of [VEHICLES_PATH, ZONES_PATH, TARIFFS_PATH]) {
      const response = await call(path);
      assert.equal(response.status, 200, `${path}: ${response.text}`);
    }
  });

  test('nothing in the catalog names a renter, a route, an invoice or a history', async () => {
    const response = await call(VEHICLES_PATH);
    for (const word of PRIVATE_WORDS) {
      assert.ok(!response.text.toLowerCase().includes(word), `the catalog published ${word}`);
    }
  });

  test('one vehicle is readable on its own and an unknown one is not', async () => {
    const catalog = await readCatalog();
    const known = await call(`${VEHICLES_PATH}/${catalog.items[0].id}`);
    assert.equal(known.status, 200, known.text);
    assert.equal(known.json.id, catalog.items[0].id);

    const unknown = await call(`${VEHICLES_PATH}/01994342-6ba7-7000-8000-ffffffffffff`);
    assert.equal(unknown.status, 404, unknown.text);
    assert.equal(unknown.json.code, 'RESOURCE_NOT_FOUND');
  });
});

describe('the service zone in PostGIS', () => {
  test('is valid WGS84 geometry the database itself accepts', () => {
    assert.equal(scalar(`SELECT bool_and(ST_IsValid(area)) FROM service_zones`), 't');
    assert.equal(scalar(`SELECT bool_and(ST_SRID(area) = 4326) FROM service_zones`), 't');
  });

  test('covers a point inside it and a point on its boundary, but not one outside', () => {
    const covered = (point) => scalar(`SELECT bool_or(ST_Covers(area, ${point})) FROM service_zones`);
    assert.equal(covered(pointAt(74.6, 42.87)), 't', 'a point inside the zone');
    assert.equal(covered(pointAt(74.55, 42.87)), 't', 'a point on the western edge');
    assert.equal(covered(pointAt(74.55, 42.84)), 't', 'a corner of the zone');
    assert.equal(covered(pointAt(74.549999, 42.87)), 'f', 'a point just outside the western edge');
    assert.equal(covered(pointAt(42.87, 74.6)), 'f', 'a swapped coordinate pair');
  });

  test('holds every demonstration vehicle', () => {
    const outside = scalar(`
      SELECT count(*) FROM vehicle_telemetry telemetry
      WHERE NOT EXISTS (
        SELECT 1 FROM service_zones zone WHERE ST_Covers(zone.area, telemetry.position)
      )`);
    assert.equal(outside, '0');
  });

  test('reports a vehicle parked outside it as outside it, beside its other reasons', async () => {
    const catalog = await readCatalog();
    const exhausted = catalog.items.find(
      (vehicle) => vehicle.status === 'unavailable' && vehicle.unavailable_reasons.includes('insufficient_energy'),
    );
    assert.ok(exhausted, 'the scenario has no exhausted vehicle');

    // Just west of the western edge, which the boundary case above shows is not covered.
    sql(`
      UPDATE vehicle_telemetry
      SET position = ${pointAt(74.549999, 42.87)}
      WHERE vehicle_id = '${exhausted.id}'`);
    try {
      const moved = await call(`${VEHICLES_PATH}/${exhausted.id}`);
      assert.equal(moved.status, 200, moved.text);
      assert.equal(moved.json.status, 'unavailable');
      assert.ok(moved.json.unavailable_reasons.includes('outside_service_zone'), moved.text);
      assert.ok(moved.json.unavailable_reasons.includes('insufficient_energy'), moved.text);
    } finally {
      restoreScenario();
    }
  });

  test('is published longitude first, as the contract writes a coordinate', async () => {
    const response = await call(ZONES_PATH);
    assert.equal(response.status, 200, response.text);
    const [[first]] = response.json.items[0].geometry.coordinates;
    const [longitude, latitude] = first;
    assert.ok(longitude > 74 && longitude < 75, `longitude was ${longitude}`);
    assert.ok(latitude > 42 && latitude < 43, `latitude was ${latitude}`);
  });
});

describe('the demonstration tariff', () => {
  test('is published in whole tyiyn under the agreed policy', async () => {
    const response = await call(TARIFFS_PATH);
    assert.equal(response.status, 200, response.text);
    const [tariff] = response.json.items;
    assert.equal(tariff.currency, 'KGS');
    assert.equal(tariff.billing_policy, 'per_mode_started_minute_v1');
    assert.equal(tariff.driving_rate_tyiyn_per_started_minute, String(DRIVING_RATE_TYIYN));
    assert.equal(tariff.paused_rate_tyiyn_per_started_minute, String(PAUSED_RATE_TYIYN));
  });
});
